core/src/main/java/com/linecorp/armeria/client/RefreshingAddressResolver.java

```diff
@@ -0,0 +1,380 @@
+                return null;
+            }
+            promise.trySuccess(new InetSocketAddress(entry.address(), port));
+            return null;
+        });
+    }
+
+    private static void handle(CompletableFuture<CacheEntry> future, Promise<InetSocketAddress> promise,
+                               int port) {
+        future.handle((entry, cause) -> {
+            if (cause != null) {
+                promise.tryFailure(cause);
+                return null;
+            }
+            entry.incrementCacheHit();
+            promise.trySuccess(new InetSocketAddress(entry.address(), port));
+            return null;
+        });
+    }
+
+    private void sendQueries(List<DnsQuestion> questions, String hostname,
+                             CompletableFuture<CacheEntry> result) {
+        final Future<List<DnsRecord>> recordsFuture = resolver.sendQueries(questions, hostname);
+        recordsFuture.addListener(f -> {
+            if (!f.isSuccess()) {
+                result.completeExceptionally(f.cause());
+                return;
+            }
+
+            @SuppressWarnings("unchecked")
+            final List<DnsRecord> records = (List<DnsRecord>) f.getNow();
+            InetAddress inetAddress = null;
+            long ttlMillis = -1;
+            try {
+                for (DnsRecord r : records) {
+                    final byte[] addrBytes = extractAddressBytes(r, logger, hostname);
+                    if (addrBytes == null) {
+                        continue;
+                    }
+                    try {
+                        inetAddress = InetAddress.getByAddress(hostname, addrBytes);
+                        ttlMillis = TimeUnit.SECONDS.toMillis(
+                                Math.max(Math.min(r.timeToLive(), maxTtl), minTtl));
+                        break;
+                    } catch (UnknownHostException e) {
+                        // Should never reach here because we already validated it in extractAddressBytes.
+                        result.completeExceptionally(
+                                new IllegalArgumentException("Invalid address: " + hostname, e));
+                        return;
+                    }
+                }
+            } finally {
+                records.forEach(ReferenceCountUtil::safeRelease);
+            }
+
+            if (inetAddress == null) {
+                result.completeExceptionally(
+                        new UnknownHostException("failed to receive DNS records for " + hostname));
+                return;
+            }
+            final CacheEntry entry = new CacheEntry(inetAddress, ttlMillis, questions);
+            result.complete(entry);
+            entry.scheduleCacheUpdate(ttlMillis);
+        });
+    }
+
+    @Override
+    protected void doResolveAll(InetSocketAddress unresolvedAddress, Promise<List<InetSocketAddress>> promise)
+            throws Exception {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
```

Comment:
Do you know when newResolver and close are called? It would help understanding the lifecycle of the refresh scheduler

Reply:
Let me add a comment. :)